feat(core): support TypeScript 4.2

crisbeto opened the pull request:
Updates the repo to TypeScript 4.2 and tslib 2.1.0.

Julien-Marcou commented:
We should wait for the TypeScript team to back port this issue https://github.com/microsoft/TypeScript/issues/43053 to the 4.2.X version of TypeScript (you can track the progress here : https://github.com/microsoft/TypeScript/pull/43180)

The deprecation warnings given by TSLint + RxJS + TypeScript 4.2 can cause a lot of CI pipelines to break

JoostK commented:
Related PR that updates ngcc to be able to deal with TS4.2 generated libraries: #41201.

JoostK commented on integration/BUILD.bazel:
```diff
@@ -97,6 +97,12 @@ INTEGRATION_TESTS = {
     },
+    "typings_test_ts42": {
+        # Special case for `typings_test_ts42` test as we want to pin
+        # `typescript` at version 4.1.x for that test and not link to the
```
```suggestion
        # `typescript` at version 4.2.x for that test and not link to the
```

JiaLiPassion reviewed (approved):
LGTM, Reviewed-for: zone.js

alan-agius4 commented:
Typically in majors we drop support for earlier TS versions. Are we going to drop support for < 4.2 in a seperate PR?

AndrewKushnir reviewed (approved):
Reviewed-for: fw-forms
Reviewed-for: fw-common

atscott reviewed (approved):
reviewed-for: fw-router

atscott reviewed (approved):
reviewed-for: language-service

IgorMinar commented on packages/compiler-cli/src/typescript_support.ts:
```diff
@@ -25,7 +25,7 @@ const MIN_TS_VERSION = '4.0.0';
  * `angularCompilerOptions.disableTypeScriptVersionCheck` config param value in g3.
  */
-const MAX_TS_VERSION = '4.2.0';
+const MAX_TS_VERSION = '4.3.0';
```
please update the MIN_TS_VERSION const to '4.2.3'

IgorMinar commented on packages/compiler-cli/package.json:
```diff
@@ -24,12 +24,12 @@
   "peerDependencies": {
     "@angular/compiler": "0.0.0-PLACEHOLDER",
-    "typescript": ">=4.0 <4.2"
+    "typescript": ">=4.0 <4.3"
```
As discussed please set the lower limit to 4.2.3. There might be a few more places where this will need to be done.

```suggestion
    "typescript": ">=4.2.3 <4.3"
```

IgorMinar commented on packages/bazel/package.json:
```diff
@@ -28,7 +28,7 @@
     "@bazel/typescript": ">=1.0.0",
     "terser": "^4.3.1",
-    "typescript": ">=4.0 <4.2",
+    "typescript": ">=4.0 <4.3",
```
```suggestion
    "typescript": ">=4.2.3 <4.3",
```

IgorMinar reviewed (approved):
as discussed offline, please increase the lower limit for typescript support to 4.2.3 and delete the 4.0 typings tests. Let's keep the 4.1 typings tests for now while g3 is being updated.

otherwise lgtm. thank you!

Reviewed-for: global-approvers

crisbeto committed:
feat(core): support TypeScript 4.2

Updates the repo to TypeScript 4.2 and tslib 2.1.0.

JoostK commented on package.json:
```diff
@@ -154,7 +154,7 @@
     "tslib": "^2.1.0",
     "tslint": "6.1.3",
-    "typescript": "~4.2.3",
+    "typescript": "^4.2.3",
```
Was changing `~` into `^` intentional? Since TS isn't following semver I'd stick to `~`.

crisbeto replied:
I did it, because we had some places using `~` and others `^` and I wanted to keep them consistent. I'll revert it.

JoostK commented on integration/BUILD.bazel:
```diff
@@ -85,12 +85,6 @@ INTEGRATION_TESTS = {
     "service-worker-schema": {},
     "side-effects": {"tags": ["no-ivy-aot"]},
     "terser": {},
-    "typings_test_ts40": {
```
Removing this here also allows removing `integration/typings_test_ts40`

crisbeto committed:
feat(core): drop support for TypeScript 4.0 and 4.1

Drops support for TypeScript 4.0 and 4.1 across the repo.
The typings check for 4.1 was kept in order to ensure that we don't break g3.

alxhub reviewed (approved):
LGTM for the compiler.

Changes for TCBs seem to be related to `ts.Printer` differences around casts, which shouldn't impact the behavior of TCBs.

AndrewKushnir commented:
[Presubmit](http://test/OCL:363321261:BASE:363321278:1615946682269:9ab293c7).

AndrewKushnir commented:
FYI, presubmit is successful for the changes in this PR.
